### DISCUSSION TITLE
Building Scalable with Vite

### DISCUSSION THREAD

**Discussion by charlieforward9** [vikejs/vike#658] [Discussion] [answered]
I am employed at a startup that has hired me to build a scalable version of their existing application. I have been doing a plethora of research to determine which tools would be the best for the job with a few factors in mind: ease of scaling resources, onboarding new developers, and optimizing user experience.

So far, I have reached this tool kit: React with Typescript for the interface, vite-plugin-ssr for middleware, and... the team wants to hook up AWS Amplify for simplified GraphQL data serving, auth, storage, etc... 

Since there are currently no resources online for this stack, I want to make sure I am making the right decision by consulting with experienced developers. 

**TL;DR - Is it safe to assume that vite-plugin-ssr integrates well with AWS amplify for all of the backend capabilities of an application that needs to be scalable?**

Thank you!

**Comment by brillout** [comment, marked as answer]
Author of vite-plugin-ssr here.

[VPS is just a server agnostic middleware](https://vite-plugin-ssr.com/integration#server-tools) so, yes, integrating with AWS amplify isn't a problem in principle.

Make sure to use the latest version of AWS Amplify because older versions had invalid ESM which required a workaround (shouldn't be needed anymore). In case you're curious this was the issue: https://github.com/aws-amplify/amplify-ui/issues/3155. Also see [this comment](https://github.com/aws-amplify/amplify-ui/issues/3155#issuecomment-1412548028) mentioning the need for another Vite workaround (seems to be still needed).

Feel free to let me know if you run into any blockers using VPS + AWS Amplify.


<br/>

> needs to be scalable

From a developer experience perspective, Vite is scalable (since it does "lazy transpiling": increasing the number of pages doesn't slow down Vite).

From a performance perspective, VPS is scalable as it does code-splitting (each page only loads what it needs).

From an architectural perspective, VPS gives you pretty much full freedom so things are scalable in that regard as well.

<br/>

Also, in case your startup has / grows to have the financial bandwidth: https://github.com/sponsors/brillout. Sponsors can have a direct impact on how VPS evolves.

You'll also get "premium" support, if that's something you'd be interested in. For example:

> ease of scaling resources, onboarding new developers, and optimizing user experience.

That's a topic I'm passionate about and I'm more than happy to discuss scaling strategies.

<br/>

I'm curious and if you don't mind me asking: how did you find out about vite-plugin-ssr?

**Reply by charlieforward9** [reply]
Rom, thank you so much for the detailed and passionate response. I have so much appreciation for open-source work. With this positive and supportive response, I will definitely move forward with VPS in our project. 

To answer your question, I have been extracting web dev knowledge from @fireship-io YouTube channel. Gotta be one of the best CS YouTubers out there.

In regard to sponsorship, I will definitely consider this as the plug-in facilitates more positive experiences down the road. Our team will compensate what is deserved, and if not the team, then me personally. 

Again, thank you for the prompt response.

**Comment by charlieforward9** [comment]
@brillout On this note, I notice the file structure is much different than a typical Vite project. When configuring Amplify, I am asked to specify the [source directory path](https://docs.amplify.aws/start/getting-started/setup/q/integration/react/#initialize-a-new-backend), however there is no `src/` folder with this setup. I see the option to [add an optional src directory](https://vite-plugin-ssr.com/file-structure#src-pages), but I would like to make sure before moving forward. 


<img width="135" alt="Screen Shot 2023-02-21 at 10 40 08 AM" src="https://user-images.githubusercontent.com/62311337/220390907-d7049f5e-dfbb-47e7-a933-1de14decb212.png">

**Reply by brillout** [reply]
Yes you can do that, I'm not aware of any issues with moving files into `src/`.

**Reply by charlieforward9** [reply]
Fantastic, and there is no issue with `server/` being outside of `src/`?

**Reply by brillout** [reply]
Not that I'm aware of.

**Reply by charlieforward9** [reply]
Guess I'll find out. Thank you!

**Reply by fortezhuo** [reply]
I am still playing with VPS. and I found something weird while build, and 

Using src folder
<img width="373" alt="image" src="https://user-images.githubusercontent.com/9128724/222305136-98cb4726-3654-4986-a33f-1a56ae1304f0.png">

Without src folder
<img width="373" alt="image" src="https://user-images.githubusercontent.com/9128724/222305402-4d6a47d3-d737-4563-9c23-5b4aa1235232.png">

We will find, the build structure changes from build/client/assets to build/client/assets/src

How to reproduce 
1. Clone this repo : https://github.com/fortezhuo/fastify-vite
2. Create src, and move pages, renderer, server into src folder
3. Add baseUrl : "src" on tsconfig.json
4. Change package.json  "start": "npm run build && cross-env NODE_ENV=production ts-node --esm ./src/server/index.ts"

I try to change server/index.ts to make server production running but no luck

```
const root = `${__dirname}/..`

startServer()

async function startServer() {
  const app = fastify()

  await app.register(middie)
  await app.register(compress)

  if (isProduction) {
    const distPath = path.join(root, '/build/client/assets')
   // try to play >> const distPath = path.join(root, '/build/client/assets/src') // not working

````

**Reply by brillout** [reply]
@fortezhuo Does it also happen without using Fastify? E.g. with one of the Express.js boilerplate `$ pnpm create vite-plugin-ssr`?

**Reply by fortezhuo** [reply]
Unfortunately, I don't use express js on this apps. So I can't confirm it also happend or not.

**Reply by brillout** [reply]
Do you mind creating a minimal reproduction on top of a boilerplate? That would be helpful. The thing is that I suspect that it's a user land problem.

**Reply by fortezhuo** [reply]
Here : https://github.com/fortezhuo/vite-plugin-ssr And on express js. It run properly in src folder.

**Reply by brillout** [reply]
> It run properly in src folder.

So the bug only happens with your Fastify example, correct?

**Reply by fortezhuo** [reply]
Yes. I confirmed it also works on Fastify after I investigated. Thanks @brillout for your time
